### DISCUSSION TITLE
Is having multiple for loops for iterating through the components in 'view' or a 'group', wrong thing to do?

### DISCUSSION THREAD

**Discussion by Doctor-Foxling** [skypjack/entt#681] [Q&A] [answered]
So I'm fairly new to entt and was using entt mostly through trial and error.

So In my use case, I have something like this:

`
 
           if (mainCamera)
		{
			Renderer2D::BeginScene(mainCamera->GetProjection(), cameraTransform);

			auto group = m_Registry.group<TransformComponent>(entt::get<SpriteRendererComponent, CubeRendererComponent>);
			for (auto entity : group)
			{
				auto [transform, sprite] = group.get<TransformComponent, SpriteRendererComponent>(entity);

				Renderer2D::DrawQuad(transform.GetTransform(), sprite.Color);
			}

			Renderer2D::EndScene();

			Renderer3D::BeginScene(mainCamera->GetProjection(), cameraTransform);

			for (auto entity : group)
			{
				
				auto [transform, cube] = group.get<TransformComponent, CubeRendererComponent>(entity);

				Renderer3D::DrawCube(transform.GetTransform(), cube.Color);
			}

			Renderer3D::EndScene();

		}`

Where I am iterating through a group twice, since I need certain things to be sandwiched between the begin/end scenes of the separate renderers.

I am not getting anything rendered with code above. It works fine for either 'cube' or 'sprite' if I do (the following) where I just extract one component but not both at the same time:

`

                  auto group = m_Registry.group<TransformComponent>(entt::get<SpriteRendererComponent>);
			for (auto entity : group)
			{
				auto [transform, sprite] = group.get<TransformComponent, SpriteRendererComponent>(entity);

				Renderer2D::DrawQuad(transform.GetTransform(), sprite.Color);
			}`

Originally I was creating 2 differrent groups (like shown below) in different scopes but that gave me a weird runtime error.

`
                       
                     auto group = m_Registry.group<TransformComponent>(entt::get<SpriteRendererComponent>);

			auto group2 = m_Registry.group<TransformComponent>(entt::get<CubeRendererComponent>);`

I guess my questions mainly boil down to 3 things:

- Why can't I loop through a group twice to extract different components (if that is in fact an issue).
- Why can't I have two separate groups that partially owns the same component and fully-owns the second component
- What's the alternative here

Thanks.

**Comment by Innokentiy-Alaytsev** [comment]
Do your entities have both components at the same time? If not then having the group constructed the way you do in the first snippet is incorrect - `entt::get<...>` parameter suggests entities have both components at the same time.

**Reply by Doctor-Foxling** [reply]
Hi, sorry for the late reply. Yes, I can't believe I didn't realize that, Thank you.

That does however, leaves the question as to what the alternative is.

Creating two different groups (within different scopes) that partially own the 'transformComponent' and fully own the 'SpriteRendererComponent' and 'CubeRendererComponent' in the respective groups, gives me the following error:

`Assertion failed: std::all_of(groups.cbegin(), groups.cend(), [size](const auto &gdata) { const auto overlapping = (0u + ... + gdata.owned(type_hash<std::decay_t<Owned>>::value())); const auto sz = overlapping + (0u + ... + gdata.get(type_hash<std::decay_t<Get>>::value())) + (0u + ... + gdata.exclude(type_hash<Exclude>::value())); return !overlapping || ((sz == size) || (sz == gdata.size)); }), file C:\Devz\BT_3D\Becketron\vendor\entt\include\entt.hpp, line 15934`

**Comment by skypjack** [comment, marked as answer]
> Why can't I loop through a group twice to extract different components (if that is in fact an issue).

You can. There is nothing wrong in it. Though, without a reproducible example, it's impossible to tell you what's weong in what you're doing.

> Why can't I have two separate groups that partially owns the same component and fully-owns the second component

It's the other way around. You cannot own the same component:
```cpp
auto group = m_Registry.group<SpriteRendererComponent>(entt::get<TransformComponent>);
auto group2 = m_Registry.group<CubeRendererComponent>(entt::get<TransformComponent>);
```
Or even:
```cpp
auto group = m_Registry.group<SpriteRendererComponent, TransformComponent>();
auto group2 = m_Registry.group<CubeRendererComponent>(entt::get<TransformComponent>);
```
And so on.

> What's the alternative here

An alternative to what exactly? You can use views and probably you should do it until you measure and find that there is a bottleneck to justify the use of a group. Other than this, as I said above, withou a self-contained snippet that explains what you're trying to do it's hard to tell what an alternative is for that. 🤷‍♂️

**Reply by Doctor-Foxling** [reply]
Hi Michele, Thanks for the reply.

So here's what I was initially trying to do:

`             
             
            if (mainCamera)
		{
			Renderer2D::BeginScene(mainCamera->GetProjection(), cameraTransform);


			auto group = m_Registry.group<TransformComponent>(entt::get<SpriteRendererComponent>);
			for (auto entity : group)
			{
				auto [transform, sprite] = group.get<TransformComponent, SpriteRendererComponent>(entity);

				Renderer2D::DrawQuad(transform.GetTransform(), sprite.Color);
			}

			Renderer2D::EndScene();
		}

		if (mainCamera)
		{
			Renderer3D::BeginScene(mainCamera->GetProjection(), cameraTransform);



			auto group2 = m_Registry.group<TransformComponent>(entt::get<CubeRendererComponent>);
			for (auto entity : group2)
			{
				auto [transform, cube] = group2.get<TransformComponent, CubeRendererComponent>(entity);

				Renderer3D::DrawCube(transform.GetTransform(), cube.Color);
			}

			Renderer3D::EndScene();
		}
	}`


And, when I run it I get the following error:

`Assertion failed: std::all_of(groups.cbegin(), groups.cend(), [size](const auto &gdata) { const auto overlapping = (0u + ... + gdata.owned(type_hash<std::decay_t<Owned>>::value())); const auto sz = overlapping + (0u + ... + gdata.get(type_hash<std::decay_t<Get>>::value())) + (0u + ... + gdata.exclude(type_hash<Exclude>::value())); return !overlapping || ((sz == size) || (sz == gdata.size)); }), file C:\Devz\BT_3D\Becketron\vendor\entt\include\entt.hpp, line 15934`

I have made sure that it's not a problem with my code API, which works without implementing ECS.

Thanks.

**Reply by skypjack** [reply]
You've two groups that own `TransformComponent` and this isn't allowed. Hence the error. Look at my previous message.

**Reply by Doctor-Foxling** [reply]
Thanks. It seems I misunderstood the concept of owning 😳 . I thought the syntax meant they partially owned the transform component.

For now, using 'view' seems to have fixed all my problems and I am going to stick to that until I get a better understanding of the system.
